enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,324 @@
+
+#### Story 1
+
+As a customer with OpenShift running in AWS and connected to my data center by VPN, I want OpenShift DNS to resolve
+name queries for our other internal devices using the DNS servers in our data center.
+
+### Implementation Details/Notes/Constraints
+
+If `Forwarders` consists of more than one `Forwarder`, longest suffix match will be used to determine the `Forwarder`.
+For example, if there are two `Forwarders`, one for foo.com and one for a.foo.com, and the query is for www.a.foo.com,
+it will be routed to the a.foo.com `Forwarder`.
+
+A maximum of 15 `Nameservers` is allowed per `Forwarder`.
+
+The cluster-dns-operator will generate the `forward` portion of `configmap/dns-default` based on `Forwarders` of
+`dnses.operator.openshift.io` instead of `forward` being statically defined. To achieve this:
+
+1. The [default](https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml) ConfigMap
+asset and associated code from the [manifests pkg](https://github.com/openshift/cluster-dns-operator/blob/master//pkg/manifests/manifests.go#L70:6)
+should be removed.
+
+2. The [desiredDNSConfigMap](https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/controller_dns_configmap.go)
+function must be modified to create a ConfigMap type with a `forward` configuration based on `Forwarder`.
+If `Forwarder` is not present or an invalid `Forwarder` is provided, the ConfigMap will contain
+`forward . /etc/resolv.conf`. Otherwise, desiredDNSConfigMap will use the provided `Forwarder` to construct the
+forwarding configuration and use `/etc/resolv.conf` as a resolver of last resort. For example:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  forwarders:
+  - domain: foo.com
+    nameServers:
+      - 1.2.3.4
+      - 5.6.7.8:5353
+    policy: RoundRobin
+  - domain: bar.com
+    nameServers:
+      - 4.3.2.1
+      - 8.7.6.5:5353
+    policy: Sequential
+```
+
+The above `DNS` will produce the following `ConfigMap`:
+```yaml
+apiVersion: v1
+data:
+  Corefile: |
+    foo.com:5353 {
+        forward . 1.2.3.4 5.6.7.8:5353
+        policy round_robin
+    }
+    bar.com:5353 {
+        forward . 4.3.2.1 8.7.6.5:5353
+        policy sequential
+    }
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf {
+            policy sequential
+        }
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+```
+#### Updating
+Changes to `Forwarder` will trigger a [rolling update](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#performing-a-rolling-update)
+of the CoreDNS DaemonSet.
```

Comment:
What happens on a given node when that node's CoreDNS is restarting?  Do we cordon/drain/restart/uncordon?  Or do pods on that node just have to live with the lack of DNS resolution during the restart?

Reply:
> Or do pods on that node just have to live with the lack of DNS resolution during the restart?

Queries go to a Kubernetes service, so they'll get forwarded to a CoreDNS pod on another node.   We use a daemonset so each node has a local pod for performance, resilience, and scalability, but having a local pod isn't strictly needed.